### DISCUSSION TITLE
Roadmap item - Release workflows (build -> package -> publish)

### DISCUSSION THREAD

**Discussion by harlequin** [moonrepo/moon#1720] [Ideas]
Hey @milesj, @rhuanbarreto

I have some ideas about the mentioned **Release Workflows**, additional this can handle the #1698 item from @rhuanbarreto.

When we are adapting the project types to a list of

- run
- build
- test
- package - `new one`
- deploy - `new one`
- release - `new one`

And adapt the cli in this way

Command | Filtered Types | Use Case
--- | --- | ---
`moon check` | `build` and `test` | Local execution
`moon ci` | `build`, `test` `package`, `deploy` | During CI/Pipeline runs in GitLab or Azure (can be even used to validate Pull/Merge Request)
`moon release` | `build`, `test`, `package`, `deploy` and `release` | During CI/Pipeline runs in GitLab or Azure (on Tag or onMain)

Intention is here really to make a difference from the "CI/CD" part and "Releases"

The most important definition would be the separation of `deploy` and `release`.
We in our company heavily `deploying` software artifacts from the software department which will send to a foreign test department.

After a while they came back with the results, and if everything works fine
We will `release` an artifact via a tag

**Comment by milesj** [comment]
I'm not sure I'm sold yet on adding more task types. Bazel has 3 types (I copied them for moon) and is able to accomplish all required workflows. To me, the `package` type is simply `build`, while `deploy` and `release` are just `run`.

The original idea behind release workflows in moon was that it was tied to the toolchain. For example, if you're on the node/bun toolchain, a release would be publishing a package to npm. If the rust toolchain, publishing a crate to crates.io. If on deno, publishing to jsr. So on and so forth. The release workflow wouldn't be a custom thing unique to a user, simply use tasks, name them accordingly, and run them when you need to.

Furthermore, the release workflow would comprise of these steps:

- Create a changeset (version bump + message) in a PR
- When ready, apply all changesets
- Generate changelog
- Bump version in manifest
- Release/publish to toolchain registry

With that said, I do have plans to add a `moon cd` command, that would pair with a `runInCD` option, similarly to how CI works. The gist is that `moon ci` would run in branches, while `moon cd` would run on master once things are good.

**Reply by rhuanbarreto** [reply]
I agree with @milesj. moon cd looks good to me also. Before this moon cd command maybe we could have a more powerful way of filtering the tasks we want to run in the graph when you call `moon run`? Not sure if this could be a middle step before we actually go to a new command.

**Reply by milesj** [reply]
@rhuanbarreto There's `moon run --query`, does that provide enough filtering?
